### ISSUE TITLE
Sudden Shutdown

### ISSUE THREAD

**Issue by CellBro** [X-LANCE/SLAM-LLM#243]
### System Info

[HAMI-core Msg(302222:140174380133248:libvgpu.c:852)]: Initialized
PyTorch version: 2.7.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.23 | packaged by conda-forge | (main, Jun  4 2025, 17:57:12)  [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.14.0-570.24.1.el9_6.x86_64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3
GPU 2: NVIDIA H100 80GB HBM3
GPU 3: NVIDIA H100 80GB HBM3

Nvidia driver version: 570.158.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.8.0
/usr/lib/x86_64-linux-

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### 🐛 Describe the bug

When using your large-scale industrial training framework to train an ASR model on the WenetSpeech dataset with four H100 GPUs, I consistently encounter training stopping at step=6447. At this point, **the VRAM usage on GPU3 drops significantly while its utilization remains at 100%(Using Deepspeed,however,when using DDP,all GPUs'utilization remains at 100%,but it just suck there )**. The other GPUs are waiting for GPU3 to finish running. After a period, **an NCCL timeout occurs**(No NCCL error occurs), causing the training to be interrupted. This issue **only occurs when using DDP (Distributed Data Parallel) and DeepSpeed**.The framework works perfectly with simple DP.What could possibly go wrong🧐

### Error logs

Training Epoch: 3/5, step 6446/ completed (loss: 2.0974676609039307, acc: 0.699999988079071): : 6447it [24:34, 4.37it/s]
[rank1]:[E925 13:51:57.944323916 ProcessGroupGloo.cpp:145] Rank 1 successfully reached monitoredBarrier, but received errors while waiting for send/recv from rank 0. Please check rank 0 logs for faulty rank.
[2025-09-25 13:51:57][root][INFO] - Detected uneven workload distribution. Break current worker to manually join all workers, world_size 4, current rank 1, current local_rank 1

[rank2]:[E925 13:51:57.946590927 ProcessGroupGloo.cpp:145] Rank 2 successfully reached monitoredBarrier, but received errors while waiting for send/recv from rank 0. Please check rank 0 logs for faulty rank.
[2025-09-25 13:51:57][root][INFO] - Detected uneven workload distribution. Break current worker to manually join all workers, world_size 4, current rank 2, current local_rank 2

### Expected behavior

The program should continue trainning

**Comment by CellBro** [comment]
may have common with https://github.com/X-LANCE/SLAM-LLM/issues/109#issue-2383004130

**Comment by CellBro** [comment]
First and foremost, I want to extend **my deepest gratitude to the SLAM-LLM team❤️**—your open-source tools and code didn’t just solve my technical challenges; they also made me feel the warmth of community collaboration. Your dedication and professionalism are truly admirable!

Now that **my issue has been successfully resolved,** I’ve decided to close this issue. To help fellow developers avoid the same pitfalls, **I’ve compiled my key solutions and lessons learned below for your reference🔎:**

1. **Environment Setup: Workaround for Docker Restrictions**
Due to network limitations preventing me from using the team’s Docker image directly, I set up the environment from scratch. Here are the core configurations (tested and working):
**Software Versions:** Python 3.9.23 | NCCL 2.21.5 | DeepSpeed 0.18.0 | PyTorch 2.7.1 | PyAudio 2.7.1
**Server Dependencies:** CUDA 11.8; note that a recent version of HAMi is required (older versions cause NCCL initialization failures).
2. **DeepSpeed Deadlock Issue: Debugging & Temporary Fix**
The core problem I faced was process synchronization failure caused by Iterable Dataset—discrepancies in the number of batches across processes led some processes to finish the epoch early, while others waited indefinitely for backpropagation, resulting in GPU hangs (manifested as the training getting “stuck” after an epoch ends).

**🤯The debugging process was winding:**

Initially, I mistook it for an NCCL communication issue (due to limited log information from NCCL).
It wasn’t until I switched to the Gloo backend that I identified the root cause—process deadlock (unsynchronized epoch progress across processes).
Upon checking, **this issue still exists in the current speech_dataset_large.py.** 🧐

**🩹My temporary fix:**

**When real data runs out, loop to generate “dummy data” marked with “is_end” and use “all_gather” to synchronize all processes, forcing the epoch to terminate.** This method only requires a small number of audio clips to stabilize training—it’s not the most elegant, but it solves the problem quickly.

**Debugging Tips for Future Reference：**

- **Test with Prime-Numbered Samples:** Use a prime number of samples (e.g., 3, 7) for testing—this makes it easier to expose batch count discrepancies across processes.

- **Gloo for Deadlock Debugging:** When facing GPU hangs, switch to the Gloo backend first—it provides clearer insights into synchronization issues.

- **Memory Monitoring Observation:** SLAM-LLM’s built-in memory monitor seems to affect stability (not fully verified). You might try the official memory_monitor API as an alternative.

Once again, thank you for your open-source contributions—your work has benefited many researchers! 

**Wish the project continued success!🥰**
